lib/no_brainer/document/types/array.rb

```diff
@@ -0,0 +1,89 @@
+require 'no_brainer/document'
+require 'active_support/core_ext/array/wrap'
+
+module NoBrainer
+  class Array
+    # delegate casting methods to each array element, if defined
+    %i(nobrainer_cast_user_to_model nobrainer_cast_model_to_db nobrainer_cast_db_to_model).each do |method|
+      redefine_singleton_method(method) do |values|
+        ::Array.wrap(values).map do |value|
+          if value.class.respond_to?(method)
+            value.class.__send__(method, value)
+          else
+            value
+          end
+        end
+      end
+    end
+
+    # convenience method to create a TypedArray
+    def self.of(object_type = nil, &object_type_proc)
+      NoBrainer::TypedArray.of(object_type, &object_type_proc)
+    end
+  end
+
+  class TypedArray < ::Array
+    def self.of(object_type = nil, &object_type_proc)
+      object_type ||= object_type_proc
+      unless object_type
+        raise ArgumentError, "Expected either Object or block"
+      end
+
+      array_type = ::Class.new(TypedArray) do
+        define_singleton_method(:object_type,
+          if object_type.respond_to?(:call)
```

Comment:
This seems a bit complicated. Is there a better way?

Reply:
I removed the lazy type resolution, is it was incomplete and probably wouldn't be used anyway.